### PR TITLE
Cxx11 code updates phase2

### PR DESCRIPTION
**Pull request by hjmjohnson**
More aggressive C++11 conversions, remove unnecessary compile time testing and conditional compilation paths that are always known values under C++11.  

- Removed backward compatibility to pre cmake 3.0 configurations since 3.10.2 is the minimum required.
- Removed code that had been deprecated since 2004.
- Replace aliased typed with their C++11 standard names.   Improved the VCL_ModernizeNaming.py.
- Moved C++ headers to the top of each header file to ensure they are only included once.
- Use range-based loops for common idioms and auto keyword to improve readability when the type can be trivially deduced.
- Removed unused variables and typedefs, initialization order of member variables in constructors to remove 1000's of compiler warnings (This also uncovered a few potential bugs that were fixed).

### CODE REVIEW

**Comment by hjmjohnson** [comment]
@dzenanz @thewtex @vxl/developers This set of changes provides what I consider to be a taggable version of VXL 2.0.0 (rc1).

All of the contrib directories build (with static libraries), and ITK with USE_SYSTEM_VXL builds against this release by requiring find_package(VXL 2.0.0)

File: contrib/brl/bbas/bdgl/bdgl_curve_algs.cxx [hunk 1]
```diff
@@ -570,19 +570,18 @@ bdgl_curve_algs::match_intersection(vdgl_digital_curve_sptr const& dc,
   vgl_homg_point_2d<double> rph(ref_point.x(), ref_point.y());
   double dist = 1e10;
   bool found_valid_intersection = false;
-  for (std::vector<double>::iterator iit = indices.begin();
-       iit != indices.end(); iit++)
+  for (double & indice : indices)
```
Comment (dzenanz) [inline contrib/brl/bbas/bdgl/bdgl_curve_algs.cxx]: If it is not too much of a problem, rename `indice` into `index` as it appears a lot.

Also this commit's name (`STYLE: Use reange-baed loops from C++11`) has spelling errors and commit message has a cut-off sentence.

File: contrib/brl/bbas/bpgl/algo/bpgl_bundle_rolling_shutter_adjust.cxx [hunk 2]
```diff
@@ -136,7 +136,6 @@ bpgl_bundle_rolling_shutter_adj_lsqr::f(vnl_vector<double> const& a,
                                         vnl_vector<double> const& b,
                                         vnl_vector<double>& e)
 {
-  typedef vnl_crs_index::sparse_vector::iterator sv_itr;
```
Comment (dzenanz) [inline contrib/brl/bbas/bpgl/algo/bpgl_bundle_rolling_shutter_adjust.cxx]: Commit name misspelling `usued`->`unused`.

File: config/cmake/config/VXLIntrospectionConfig.cmake [hunk 3]
```diff
@@ -656,21 +655,6 @@ endif()
 
 # Check C++ <cmath> first, where the C++11 standard says these must be.
 include(${CMAKE_CURRENT_LIST_DIR}/CheckCXXExpressionCompiles.cmake) ## From VTK
-check_cxx_expression_compiles("std::isnan(0.0)" "cmath" VXL_HAS_STD_ISNAN)
```
Comment (dzenanz) [inline config/cmake/config/VXLIntrospectionConfig.cmake]: These removals should speed up initial configure time. Yeah!

File: contrib/brl/bbas/bgui3d/bgui3d_tableau.h [hunk 4]
```diff
@@ -12,11 +12,10 @@
 //   <none yet>
 // \endverbatim
 
-#include <iostream>
-#include <vcl_compiler.h>
 #include <iostream>
 #include <memory>
 #include <vcl_compiler.h>
+#include <vcl_compiler.h>
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply contrib/brl/bbas/bgui3d/bgui3d_tableau.h]: Fixed in 138 files for this plus many other duplicate headers.

vim $(grep -l  -Pz "(?s)#include *.(.*).*#include *.\1." $(find . -name "*.cxx" -o -name "*.hxx" -o -name "*.h" ) )

File: vcl/CMakeLists.txt [hunk 5]
```diff
@@ -14,89 +14,75 @@ endif()
...
-        cxx_unrestricted_unions                 # Unrestricted unions, as defined in N2544.
-        cxx_user_literals                       # User-defined literals, as defined in N2765.
-        cxx_variable_templates                  # Variable templates, as defined in N3651.
-        cxx_variadic_macros                     # Variadic macros, as defined in N1653.
-        cxx_variadic_templates                  # Variadic templates, as defined in N2242.
-        cxx_template_template_parameters        # Template template parameters, as defined in ISO/IEC 14882:1998.
-  )
-endif()
+## Dynamically create compiler detection header
+set(VCL_COMPILER_DETECTION_HEADER "${CMAKE_CURRENT_BINARY_DIR}/vcl_compiler_detection.h")
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply vcl/CMakeLists.txt]: I think it should make configuration faster.

File: contrib/brl/bbas/bvgl/bvgl_grid_index_3d.h [hunk 6]
```diff
@@ -8,13 +8,12 @@
...
 //
 #include <iostream>
-#include <vcl_compiler.h>
-#include <iostream>
 #include <fstream>
 #include <vector>
 #include <string>
 #include <map>
 #include <vcl_compiler.h>
+#include <vcl_compiler.h>
```
[the remark this answers is not in the document]
Comment (hjmjohnson) [reply contrib/brl/bbas/bvgl/bvgl_grid_index_3d.h]: fixed.

File: contrib/brl/bbas/bpgl/algo/bpgl_bundle_rolling_shutter_adjust.cxx (hunk 2 above)
Comment (hjmjohnson) [reply contrib/brl/bbas/bpgl/algo/bpgl_bundle_rolling_shutter_adjust.cxx]: done in all 4 files

File: contrib/brl/bbas/bdgl/bdgl_curve_algs.cxx (hunk 1 above)
Comment (hjmjohnson) [reply contrib/brl/bbas/bdgl/bdgl_curve_algs.cxx]: done in all 4 files

File: config/cmake/config/VXLIntrospectionConfig.cmake (hunk 3 above)
Comment (hjmjohnson) [reply config/cmake/config/VXLIntrospectionConfig.cmake]: I hope so.  There are many more config time checks that are now meaningless.  Will remove as time permits.

**Pull request merged**